commit 3bb054db526aa07dcd88fd563dd705fabb35472f
Author: sewen <stephan.ewen@tu-berlin.de>
Date:   Wed Mar 23 20:00:03 2011 +0100

    Sorter closing is synchronized.

diff --git a/pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/sort/UnilateralSortMerger.java b/pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/sort/UnilateralSortMerger.java
index 07f8eb1..f565e9e 100644
--- a/pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/sort/UnilateralSortMerger.java
+++ b/pact/pact-runtime/src/main/java/eu/stratosphere/pact/runtime/sort/UnilateralSortMerger.java
@@ -401,7 +401,7 @@ public class UnilateralSortMerger<K extends Key, V extends Value> implements Sor
 	 * @see java.io.Closeable#close()
 	 */
 	@Override
-	public void close()
+	public synchronized void close()
 	{
 		// check if the sorter has been closed before
 		if (this.closed) {
