commit 22d79a501d85a1a0d6e22859c335fa09c0a11c9b
Author: Trustin Lee <trustin@gmail.com>
Date:   Fri Sep 26 04:31:41 2008 +0000

    Replaced ConcurrentLinkedQueue with LinkedBlockingQueue in NioSocketChannel

diff --git a/src/main/java/org/jboss/netty/channel/socket/nio/NioSocketChannel.java b/src/main/java/org/jboss/netty/channel/socket/nio/NioSocketChannel.java
index 7a3c818..92f39be 100644
--- a/src/main/java/org/jboss/netty/channel/socket/nio/NioSocketChannel.java
+++ b/src/main/java/org/jboss/netty/channel/socket/nio/NioSocketChannel.java
@@ -26,7 +26,7 @@ import java.net.InetSocketAddress;
 import java.net.SocketAddress;
 import java.nio.channels.SocketChannel;
 import java.util.Queue;
-import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.LinkedBlockingQueue;
 import java.util.concurrent.atomic.AtomicBoolean;
 
 import org.jboss.netty.channel.AbstractChannel;
@@ -52,8 +52,7 @@ abstract class NioSocketChannel extends AbstractChannel
 
     final AtomicBoolean writeTaskInTaskQueue = new AtomicBoolean();
     final Runnable writeTask = new WriteTask();
-    final Queue<MessageEvent> writeBuffer =
-        new ConcurrentLinkedQueue<MessageEvent>();
+    final Queue<MessageEvent> writeBuffer = new LinkedBlockingQueue<MessageEvent>();
     MessageEvent currentWriteEvent;
     int currentWriteIndex;
 
