commit ee019d344e56a99ea787e55738a19b2c3c615120
Author: Trustin Lee <trustin@gmail.com>
Date:   Sun Jul 8 22:44:15 2012 +0900

    Fix potential NPE in AioEventLoop / Always notify flush futures

diff --git a/transport/src/main/java/io/netty/channel/socket/aio/AioEventLoop.java b/transport/src/main/java/io/netty/channel/socket/aio/AioEventLoop.java
index bd7503b..1833554 100644
--- a/transport/src/main/java/io/netty/channel/socket/aio/AioEventLoop.java
+++ b/transport/src/main/java/io/netty/channel/socket/aio/AioEventLoop.java
@@ -73,7 +73,7 @@ public class AioEventLoop extends MultithreadEventLoop {
         if (l.isShutdown()) {
             command.run();
         } else {
-            ch.eventLoop().execute(command);
+            l.execute(command);
         }
     }
 
diff --git a/transport/src/main/java/io/netty/channel/socket/aio/AioSocketChannel.java b/transport/src/main/java/io/netty/channel/socket/aio/AioSocketChannel.java
index ea6bc0c..c79e3da 100644
--- a/transport/src/main/java/io/netty/channel/socket/aio/AioSocketChannel.java
+++ b/transport/src/main/java/io/netty/channel/socket/aio/AioSocketChannel.java
@@ -165,6 +165,7 @@ public class AioSocketChannel extends AbstractAioChannel implements SocketChanne
     @Override
     protected void doFlushByteBuffer(ByteBuf buf) throws Exception {
         if (!buf.readable()) {
+            notifyFlushFutures();
             return;
         }
 
