commit 8e55bbd41725cd8c7f29a9e41993a22777ebdacf
Author: Greg Hogan <code@greghogan.com>
Date:   Fri Feb 12 08:38:05 2016 -0500

    [FLINK-3393] [core] ExternalProcessRunner wait to finish copying error stream

diff --git a/flink-core/src/main/java/org/apache/flink/util/ExternalProcessRunner.java b/flink-core/src/main/java/org/apache/flink/util/ExternalProcessRunner.java
index 8e4725c..b7e388c 100644
--- a/flink-core/src/main/java/org/apache/flink/util/ExternalProcessRunner.java
+++ b/flink-core/src/main/java/org/apache/flink/util/ExternalProcessRunner.java
@@ -41,6 +41,8 @@ public class ExternalProcessRunner {
 
 	private final Process process;
 
+	private final Thread pipeForwarder;
+
 	final StringWriter errorOutput = new StringWriter();
 
 	/**
@@ -63,7 +65,7 @@ public class ExternalProcessRunner {
 
 		process = new ProcessBuilder(commandList).start();
 
-		new PipeForwarder(process.getErrorStream(), errorOutput);
+		pipeForwarder = new PipeForwarder(process.getErrorStream(), errorOutput);
 	}
 
 	/**
@@ -83,6 +85,9 @@ public class ExternalProcessRunner {
 		try {
 			int returnCode = process.waitFor();
 
+			// wait to finish copying standard error stream
+			pipeForwarder.join();
+
 			if (returnCode != 0) {
 				// determine whether we failed because of a ClassNotFoundException and forward that
 				if (getErrorOutput().toString().contains("Error: Could not find or load main class " + entryPointClassName)) {
