commit 0cb3541b53f5e5673ab3fe16702fb98c522a0ce2
Author: Norman Maurer <nmaurer@redhat.com>
Date:   Mon Jul 29 08:09:52 2013 +0200

    [#1669] Correctly notify the ChannelPromise of delayed writes

diff --git a/handler/src/main/java/io/netty/handler/traffic/AbstractTrafficShapingHandler.java b/handler/src/main/java/io/netty/handler/traffic/AbstractTrafficShapingHandler.java
index 35343e8..8a48d24 100644
--- a/handler/src/main/java/io/netty/handler/traffic/AbstractTrafficShapingHandler.java
+++ b/handler/src/main/java/io/netty/handler/traffic/AbstractTrafficShapingHandler.java
@@ -277,7 +277,7 @@ public abstract class AbstractTrafficShapingHandler extends ChannelDuplexHandler
     }
 
     @Override
-    public void write(final ChannelHandlerContext ctx, final Object msg, ChannelPromise promise)
+    public void write(final ChannelHandlerContext ctx, final Object msg, final ChannelPromise promise)
             throws Exception {
         long curtime = System.currentTimeMillis();
         long size = calculateSize(msg);
@@ -297,7 +297,7 @@ public abstract class AbstractTrafficShapingHandler extends ChannelDuplexHandler
                 ctx.executor().schedule(new Runnable() {
                     @Override
                     public void run() {
-                        ctx.write(msg);
+                        ctx.write(msg, promise);
                     }
                 }, wait, TimeUnit.MILLISECONDS);
                 return;
