commit 9d3828bbe91d7bc7aad6e34c719c4591593be84d
Author: Norman Maurer <norman@apache.org>
Date:   Wed Nov 30 20:14:57 2011 +0100

    Use LinkedTransferQueue for OMATPE

diff --git a/src/main/java/org/jboss/netty/handler/execution/OrderedMemoryAwareThreadPoolExecutor.java b/src/main/java/org/jboss/netty/handler/execution/OrderedMemoryAwareThreadPoolExecutor.java
index ede4134..c6949ef 100644
--- a/src/main/java/org/jboss/netty/handler/execution/OrderedMemoryAwareThreadPoolExecutor.java
+++ b/src/main/java/org/jboss/netty/handler/execution/OrderedMemoryAwareThreadPoolExecutor.java
@@ -16,9 +16,9 @@
 package org.jboss.netty.handler.execution;
 
 import java.util.IdentityHashMap;
+import java.util.Queue;
 import java.util.Set;
 import java.util.WeakHashMap;
-import java.util.concurrent.ConcurrentLinkedQueue;
 import java.util.concurrent.ConcurrentMap;
 import java.util.concurrent.Executor;
 import java.util.concurrent.ThreadFactory;
@@ -31,6 +31,7 @@ import org.jboss.netty.channel.ChannelState;
 import org.jboss.netty.channel.ChannelStateEvent;
 import org.jboss.netty.util.ObjectSizeEstimator;
 import org.jboss.netty.util.internal.ConcurrentIdentityWeakKeyHashMap;
+import org.jboss.netty.util.internal.LinkedTransferQueue;
 
 /**
  * A {@link MemoryAwareThreadPoolExecutor} which makes sure the events from the
@@ -285,7 +286,7 @@ public class OrderedMemoryAwareThreadPoolExecutor extends
     }
 
     private final class ChildExecutor implements Executor, Runnable {
-        private final ConcurrentLinkedQueue<Runnable> tasks = new ConcurrentLinkedQueue<Runnable>();
+        private final Queue<Runnable> tasks = new LinkedTransferQueue<Runnable>();
         private final AtomicBoolean isRunning = new AtomicBoolean(false);
         
         ChildExecutor() {
