commit f7e7d8f52417eac4b379df67daaa925de9a91c97
Author: Trustin Lee <trustin@gmail.com>
Date:   Sat Dec 6 14:35:13 2008 +0000

    Workaround for NPE on ExecutorService.shutdownNow()

diff --git a/src/main/java/org/jboss/netty/util/ExecutorShutdownUtil.java b/src/main/java/org/jboss/netty/util/ExecutorShutdownUtil.java
index 7eebb02..481c269 100644
--- a/src/main/java/org/jboss/netty/util/ExecutorShutdownUtil.java
+++ b/src/main/java/org/jboss/netty/util/ExecutorShutdownUtil.java
@@ -57,7 +57,12 @@ public class ExecutorShutdownUtil {
 
             ExecutorService es = (ExecutorService) e;
             for (;;) {
-                es.shutdownNow();
+                try {
+                    es.shutdownNow();
+                } catch (NullPointerException ex) {
+                    // Some JDK throws NPE here, but shouldn't.
+                }
+
                 try {
                     if (es.awaitTermination(1, TimeUnit.SECONDS)) {
                         break;
